addons/sale/tests/test_sale_order.py

```diff
@@ -446,6 +446,109 @@ def test_order_auto_lock_with_public_user(self):
+            'company_id': company.id,
+        })
+        tax_b = self.env['account.tax'].create({
+            'name': 'Tax B',
+            'type_tax_use': 'sale',
+            'amount_type': 'percent',
+            'amount': 15,
+            'tax_group_id': tax_groups[0].id,
+            'company_id': company.id,
+        })
+        tax_x = self.env['account.tax'].create({
+            'name': 'Tax X',
+            'type_tax_use': 'sale',
+            'amount_type': 'percent',
+            'amount': 20,
+            'tax_group_id': tax_groups[1].id,
+            'company_id': branch_x.id,
+        })
+        tax_xx = self.env['account.tax'].create({
+            'name': 'Tax XX',
+            'type_tax_use': 'sale',
+            'amount_type': 'percent',
+            'amount': 25,
+            'tax_group_id': tax_groups[2].id,
+            'company_id': branch_xx.id,
+        })
+        # create several products with different taxes combination
+        product_all_taxes = self.env['product.product'].create({
+            'name': 'Product all taxes',
+            'taxes_id': [Command.set((tax_a + tax_b + tax_x + tax_xx).ids)],
+        })
+        product_no_xx_tax = self.env['product.product'].create({
+            'name': 'Product no tax from XX',
+            'taxes_id': [Command.set((tax_a + tax_b + tax_x).ids)],
+        })
+        product_no_branch_tax = self.env['product.product'].create({
+            'name': 'Product no tax from branch',
+            'taxes_id': [Command.set((tax_a + tax_b).ids)],
+        })
+        product_no_tax = self.env['product.product'].create({
+            'name': 'Product no tax',
+            'taxes_id': [],
+        })
+        # create a SO from Branch XX
+        so_form = Form(self.env['sale.order'].with_company(branch_xx))
+        so_form.partner_id = self.partner
+        # add 4 SO lines with the different products:
+        # - Product all taxes           => tax from Branch XX should be set
+        # - Product no tax from XX      => tax from Branch X should be set
+        # - Product no tax from branch  => 2 taxes from parent company should be set
+        # - Product no tax              => no tax should be set
+        with so_form.order_line.new() as line:
+            line.product_id = product_all_taxes
+        with so_form.order_line.new() as line:
+            line.product_id = product_no_xx_tax
+        with so_form.order_line.new() as line:
+            line.product_id = product_no_branch_tax
+        with so_form.order_line.new() as line:
+            line.product_id = product_no_tax
+        so = so_form.save()
+        self.assertRecordValues(so.order_line, [
+            {'product_id': product_all_taxes.id, 'tax_id': tax_xx.ids},
+            {'product_id': product_no_xx_tax.id, 'tax_id': tax_x.ids},
+            {'product_id': product_no_branch_tax.id, 'tax_id': (tax_a + tax_b).ids},
+            {'product_id': product_no_tax.id, 'tax_id': []},
+        ])
```

Comment:
Even if the test is green, wouldn't it be better to also use `_filter_taxes_by_company` in `_compute_tax_id`, to be sure that all tax flows go through the same method ?

Reply:
Yep. I have added it.